sky/serve/autoscalers.py

```diff
@@ -315,3 +367,158 @@ def _get_replica_ids_to_scale_down(num_limit: int) -> List[int]:
+                num_provisioning_and_launched_spot += 1
+            else:
+                if info.status == serve_state.ReplicaStatus.READY:
+                    num_ready_ondemand += 1
+                num_provisioning_and_launched_ondemand += 1
+
+        logger.info(
+            'Number of alive spot instances: '
+            f'{num_provisioning_and_launched_spot}, '
+            f'Number of ready spot instances: {num_ready_spot}, '
+            'Number of alive on-demand instances: '
+            f' {num_provisioning_and_launched_ondemand}, '
+            f'Number of ready on-demand instances: {num_ready_ondemand}')
+
+        scaling_options: List[AutoscalerDecision] = []
+        all_replica_ids_to_scale_down: List[int] = []
+
+        # Once there is min_replicas number of ready new replicas, we will
+        # direct all traffic to them, we can scale down all old replicas.
+        all_replica_ids_to_scale_down.extend(
+            self.select_outdated_replicas_to_scale_down(replica_infos))
+
+        # Decide how many spot instances to launch.
+        num_spot_to_provision = (self.target_num_replicas -
+                                 self.base_ondemand_fallback_replicas)
+        if num_provisioning_and_launched_spot < num_spot_to_provision:
+            # Not enough spot instances, scale up.
+            num_spot_to_scale_up = (num_spot_to_provision -
+                                    num_provisioning_and_launched_spot)
+            for _ in range(num_spot_to_scale_up):
+                scaling_options.append(
+                    AutoscalerDecision(
+                        AutoscalerDecisionOperator.SCALE_UP,
+                        target=self._get_spot_resources_override_dict()))
+        elif num_provisioning_and_launched_spot > num_spot_to_provision:
+            # Too many spot instances, scale down.
+            # Get the replica to scale down with _select_replicas_to_scale_down
+            num_spot_to_scale_down = (num_provisioning_and_launched_spot -
+                                      num_spot_to_provision)
+            all_replica_ids_to_scale_down.extend(
+                RequestRateAutoscaler._select_replicas_to_scale_down(
+                    num_spot_to_scale_down,
+                    filter(lambda info: info.is_spot,
+                           latest_provisioning_and_launched_replicas)))
+
+        # Decide how many on-demand instances to launch.
+        num_ondemand_to_provision = self.base_ondemand_fallback_replicas
+        if self.dynamic_ondemand_fallback:
+            # `num_ready_spot` instead of `num_provisioning_and_launched_spot`
+            # because the provisioning spot can fail to UP due to the capacity
+            # issue, and on-demand should fill the gap between the required
+            # number of spot and ready spot.
+            num_ondemand_to_provision += (num_spot_to_provision -
+                                          num_ready_spot)
```

Comment:
Feeling like `num_ondemand_to_provision = max(self.target_num_replicas - num_ready_spot, self.base_ondemand_fallback_replicas)` is easier to understand; but current one also looks good to me.